### DISCUSSION TITLE
Support for PHP 8

### DISCUSSION THREAD

**Discussion by ShoshValley** [geodesicsolutions-community/geocore-community#85] [Q&A]
I am a huge fan of the software and have been using it on our site since at least 2012.  I would classify myself as a user of the software with limited experience customizing code, but with no real programming ability.

I could really use some help/advice.  My hosting company is forcing an upgrade to php 8 in less than a week. I am running GeoCore MAX DB Ver. 18.02.0.  It works in 7.4, but not in 8.  Is there any way to make version 18.02.0 work in php 8?

**Comment by vicos59** [comment]
@ShoshValley 

That's a tough place to be in. Depending upon how active board your is; does not give you much time to plan and execute an upgrade. Seems like they are being a little unreasonable given that PHP 7.4 will receive security updates through end of November this year.

https://www.php.net/supported-versions.php

Anyway, maybe you could contact support and see if they'll give you an extension for a few weeks.

As far as actual 8.0 support, @jonyo is one of the original developers and has been doing all of the code updates here. I suspect it can run on 8.0, but he can give a better answer (Actually, I know it can run on 8.0, but I am not certain it is 100% bug free). I know that 8.0 and beyond are his key goals. I am running it on 7.4.

If you can install a test copy of the new version 20.0.0 on your server, you could check it out for yourself. To get the latest, go here https://github.com/geodesicsolutions-community/geocore-community and click on the Green 'Code' button. The Docs/Wiki are here: https://geodesicsolutions.org/wiki/ and I believe they are up to date. Links to the install and update instructions are here: https://geodesicsolutions.org/resources.php

I'm sure @jonyo will chime in shortly if you want a better answer than the one I gave.

P.S. If you didn't know, the code to unlock the license is here: https://geodesicsolutions.org/resources.php ( See 18.02.0 Patch )

**Reply by jonyo** [reply]
> , @jonyo is one of the original developers and has been doing all of the code updates here

I'd like to add `... so far` -- I hope the community can pick up some more developers!  My goal is to get the project to a stable first release (with as much volunteer help as possible testing everything), after that I want to just be part of the project, not the sole developer.  I don't have the free time to keep it up (or to be more accurate, I'll likely burn out if I try to keep it up as the "only" developer long term).  I do this for 2 reasons, one to help folks out running the old software since I am in a position where I can help, and two because I find this kind of thing fun so it's somewhat of a hobby for me.  But I digress...

> I suspect it can run on 8.0, but he can give a better answer (Actually, I know it can run on 8.0, but I am not certain it is 100% bug free). I know that 8.0 and beyond are his key goals. I am running it on 7.4.

I have not tried recently.  I have changes that I recently just merged that will be part of beta 4, (see #84 ) that should be fixing the remaining lurking PHP 7.4 "deprecation warnings" after I found a tool that helps to scan for those types of problems so they can be fixed.  And these are things you don't even see on a production website since you normally have "PHP notices" turned off.  With those remaining deprecation warnings fixed, there is a chance it could work on PHP 8 with no bugs.  Or it could require huge changes.

What I can do is up my schedule to get it PHP 8 compatible, I was originally planning for "after version 20".  But maybe if it just requires changes here and there we can make version 20 compatible.

Just to manage expectations though - if I try it out (or you try it and report back), and it turns out it will take a lot more work to get it compatible with PHP 8, then we'll stick to the original plan of after version 20.  It will help if there are other developers that join in to work on it too though.  😉 

Once I have time to test it out I'll report back.  And likewise, if you have time to try it out per @vicos59 instructions ☝️ let us know what you find!

**Reply by vicos59** [reply]
> I hope the community can pick up some more developers!

@jonyo  If you concur, I can reach out to some of the 3rd party developers who were once listed on a page on the old Geodesics site (If I can still find it in the WayBackMachine). I spoke with one guy a few months ago, so I know there is at least one out there. Always a good way for coders to pick up work, helping out users who would prefer to pay for services to upgrade, etc.

**Reply by ShoshValley** [reply]
Thank you so much for all of the info.  I have not been happy with my hosting co. for a while, but felt stuck with no access to license to migrate.  With the code to unlock the license, I may just try migrating the site to the host I use for the rest of my site.  Then I can try the upgrade when I have more time - will definitely report back when I do.  Wish I could help with updates but I am definitely not a developer.

**Reply by vicos59** [reply]
Wish you the best! Be sure to make a full backup before making any changes. Always a good precaution.

**Reply by jonyo** [reply]
> @jonyo If you concur, I can reach out to some of the 3rd party developers who were once listed on a page on the old Geodesics site (If I can still find it in the WayBackMachine). I spoke with one guy a few months ago, so I know there is at least one out there. Always a good way for coders to pick up work, helping out users who would prefer to pay for services to upgrade, etc.

Sure that would be great!

I've mentioned it elsewhere, but this might be something sustainable, and it may be easier to explain it by telling the mini story of how it worked out:

A while back, I was taking client contract work (FYI I am now back to salary so don't do any client work - the extent I do is changes on this project that I'm doing for fun here and there).  Anyways, they are in the EU and were notified that new regulations would go into effect that would mean possibly a lot of declines if they did not upgrade the payment gateway.  So they contracted me to re-write the Stripe payment gateway almost completely to use a newer feature they have, that works with the new EU regulations.  They payed for my time because, well for one thing Geo was still around, and it made 0 sense to work on another company's licensed software for free 😆 .  I also could not have justified the amount of time and the speed at which it needed to be done unless it was paid time.  Anyways, so I made it, at the end I asked them about whether they minded if I sent the software to others that asked for it, since they were the ones to pay for the time.  They encouraged it and said they would be fine if it was made part of the software.  On their side, they needed it to continue accepting payments.  And making it part of the main software meant much better chance that it is maintained, if they kept it to themselves they would definitely be paying any time updates were required for new versions.

Sooo **what I'm trying to say is**, this might be one way for new features, and even some maintenance things like PHP8 compatibility, could get done.  If site owners need it, they can pay a developer to do it.  And if it benefits the software, it makes sense for the site owner to have the developer actually submit the changes so they become part of the software.  That way going forward the community both benefits from it, and also helps to make sure it continues to work in future versions.

**Reply by vicos59** [reply]
@jonyo OK, I pinged who I could in the developer world. Also hit up a few more users who had posted on the old company FB page years ago, just in case they might still be users. That probably about as much outreach as is possible. Looks like those old "scripts" sites which still show up on the top 10 search results for GeoCore are on cruise control and the driver is asleep. No response from either of them.  Hopefully, Googul l will do its thing and the new GeoCoreCE will make it to the top of the page.

**Reply by vicos59** [reply]
@jonyo Do you think it would be worthwhile to grab the old features pages from the WayBackMachine, clean them up and post them on the .org site? Beats rewriting all that stuff from scratch.

https://web.archive.org/web/20200525190920/https://geodesicsolutions.com/add-ons/free-add-ons.html

https://web.archive.org/web/20200512051130/https://geodesicsolutions.com/geocore-software/geocore-admin.html

https://web.archive.org/web/20200512045706/https://geodesicsolutions.com/geocore-software/geocore-themes.html

https://web.archive.org/web/20200523100023/https://geodesicsolutions.com/classified-ad-software-and-auction-features/features-list.html

**Reply by jonyo** [reply]
@vicos59 Sure, it wouldn't hurt, though we'd probably need to re-make the screenshots so it has the updated community edition name and all that.

If you do that, be sure to record if you use any specific tools (like any software that automatically makes mobile/desktop/tablet views or something) so we can easily re-create them later if needed.

**Reply by jonyo** [reply]
For the icons, something I have seen recently is [box icons](https://boxicons.com/) - entirely free icon set.  It may be a good resource for the website when we need icons, and in the software if someone wants to update the look.

That is a good resource as they provide SVG's which allow easily changing the color with CSS.

**Comment by jonyo** [comment]
So I loaded it up using PHP 8, and found out, or actually reminded myself, there are issues running the version of Smarty we are currently using in PHP8.

Since Smarty template engine is used on every page, it means the software won't work on PHP 8 right now.

There is actually an issue for it, #22 - I'll look into doing that next when I have time and see if I can get it going.

**Reply by jonyo** [reply]
Update:  It requires re-writing the Smarty template include used in system templates.  Anyone upgrading that is using custom system templates may need to adjust them for the new style, since Smarty changed things internally.  I'll have full instructions on it when it is released.

The good news is that I at least got it to load in PHP 8.1!  Which means it should work in 8.0 too since I'm not doing anything 8.1 special.

It will take more work and testing though, but I have the ground work in, so it's looking like version 20 "final" will be PHP 8 compatible!
